### ISSUE TITLE
Active Directory Authentication Library (ADAL) for .NET 5 Core CLR (dnxcore50) Projects

### ISSUE THREAD

**Issue by GuardRex** [AzureAD/azure-activedirectory-library-for-dotnet#206]
In `OnAuthorizationCodeReceived`, getting a context for Graph API this way ...
```c#
string userObjectId = notification.AuthenticationTicket.Principal.FindFirst("http://schemas.microsoft.com/identity/claims/objectidentifier").Value;
ClientCredential credential = new ClientCredential(ClientId, AppKey);
AuthenticationContext authContext = new AuthenticationContext(Authority);
AuthenticationResult result = await authContext.AcquireTokenByAuthorizationCodeAsync(notification.Code, new Uri(BaseAddress), credential, "https://graph.windows.net");
```
works with dnx451 1.0.0-beta4 clr x86 but fails for dnxcore50 1.0.0-beta6-* coreclr x64 with
```
The type initializer for 'Microsoft.IdentityModel.Clients.ActiveDirectory.Authenticator' threw an exception.
```
Values for ClientId, AppKey, Authority, notification.Code, and BaseAddres should be ok ... they weren't changed between the two runtimes. Dependencies are ...
```c#
    "Microsoft.IdentityModel.Protocol.Extensions": "2.0.0-*",
    "Microsoft.AspNet.Mvc": "6.0.0-*",
    "Microsoft.AspNet.Authentication.Cookies": "1.0.0-*",
    "Microsoft.AspNet.Diagnostics": "1.0.0-*",
    "Microsoft.AspNet.Server.IIS": "1.0.0-*",
    "Microsoft.AspNet.StaticFiles": "1.0.0-*",
    "Microsoft.AspNet.Tooling.Razor": "1.0.0-*",
    "Microsoft.IdentityModel.Clients.ActiveDirectory": "3.1.203031538-alpha",
    "Microsoft.Framework.ConfigurationModel": "1.0.0-*",
    "Microsoft.AspNet.Authentication.OpenIdConnect": "1.0.0-*",
    "Microsoft.AspNet.Identity": "3.0.0-*",
    "Microsoft.AspNet.Http": "1.0.0-*"
```

**Comment by GuardRex** [comment]
I'll update+close this question at [StackOverflow](http://stackoverflow.com/questions/30851308/the-type-initializer-for-microsoft-identitymodel-clients-activedirectory-authen) if answered first over there.

**Comment by GuardRex** [comment]
I upgraded my ActiveDirectory package to 3.3.205061641-alpha; no dice ... problem persists.

**Comment by GuardRex** [comment]
Here is some additional info on the exception:
```
System.TypeInitializationException: The type initializer for 'Microsoft.IdentityModel.Clients.ActiveDirectory.Authenticator' threw an exception. ---> System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime.Serialization.Json, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified. ---> System.IO.FileNotFoundException: Could not load the specified file.
 at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyName(AssemblyName assemblyName)
 at System.Runtime.Loader.AssemblyLoadContext.Resolve(IntPtr gchManagedAssemblyLoadContext, AssemblyName assemblyName)
 --- End of inner exception stack trace ---
 at Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticatorTemplate.CreateFromHost(String host)
 at Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticatorTemplateList..ctor()
 at Microsoft.IdentityModel.Clients.ActiveDirectory.Authenticator..cctor()
 --- End of inner exception stack trace ---
 at Microsoft.IdentityModel.Clients.ActiveDirectory.Authenticator.
```

**Comment by GuardRex** [comment]
@kpanwar @afshins @vibronet Hey guys ... how's it going. Any word on this one?

**Comment by GuardRex** [comment]
Looks like this is related to the NuGet package not being ready for `dnxcore50` per [discussion here](https://github.com/aspnet/Home/issues/674#issuecomment-112822724). I'll wait for a team member to respond before closing.

**Comment by kpanwar** [comment]
@GuardRex - I will rename aspnetcore50 to dnxcore50 in the nuget and will post back on this thread once done.

**Comment by GuardRex** [comment]
@kpanwar Thanks. And I responded by E-mail to @afshins, because he said,
> You seem to have downloaded an incorrect version of ADAL. ADAL v3 does not use WebException and SHA256Cng is only used in the desktop version of the library. The source code you need to look at is in the ‘dev’ branch and the two projects which are relevant to CoreCLR are ADAL.PCL and ADAL.PCL.CoreCLR.

So, I was asking if I'm supposed to get the source at [SymbolSource](http://www.symbolsource.org/Public/Metadata/NuGet/Project/Microsoft.IdentityModel.Clients.ActiveDirectory/3.1.203031538-alpha/Release/ASP.NetCore%2cVersion%3dv5.0/Microsoft.IdentityModel.Clients.ActiveDirectory.Platform) and compile it myself.
It seems like what you're saying is that you just need to make that framework change for the [NuGet package](https://www.nuget.org/packages/Microsoft.IdentityModel.Clients.ActiveDirectory/3.4.206191646-alpha) and it will work with `dnxcore50`. That's :+1:. I'll test the update when ur done.

**Comment by kpanwar** [comment]
@GuardRex - I released the updated v3 preview. check it out and let me know if you still have issues.

**Comment by GuardRex** [comment]
@kpanwar Thanks ... I'll try it tonight and post back the results. Thanks again!

**Comment by GuardRex** [comment]
@kpanwar I updated the ADAL reference to
```
"Microsoft.IdentityModel.Clients.ActiveDirectory": "3.5.207081303-alpha"
```
Since I first filed this issue, I've moved on to `dnx-coreclr-win-x64.1.0.0-beta6-12170`, and this is still a `dnxcore50`-only project. The full dependencies section looks like this:
```c#
  "dependencies": {
    "Microsoft.IdentityModel.Protocol.Extensions": "2.0.0-*",
    "Microsoft.AspNet.Mvc": "6.0.0-*",
    "Microsoft.AspNet.Authentication.Cookies": "1.0.0-*",
    "Microsoft.AspNet.Diagnostics": "1.0.0-*",
    "Microsoft.AspNet.Server.IIS": "1.0.0-*",
    "Microsoft.AspNet.StaticFiles": "1.0.0-*",
    "Microsoft.AspNet.Tooling.Razor": "1.0.0-*",
    "Microsoft.Framework.ConfigurationModel": "1.0.0-*",
    "Microsoft.AspNet.Authentication.OpenIdConnect": "1.0.0-*",
    "Microsoft.AspNet.Identity": "3.0.0-*",
    "Microsoft.AspNet.Http": "1.0.0-*",
    "Microsoft.IdentityModel.Clients.ActiveDirectory": "3.5.207081303-alpha"
  },
```
It's throwing a new exception after AAD gets hit when I try to get a token for Graph API at Line 98 of my Startup.cs, which is:
```c#
AuthenticationResult result = await authContext.AcquireTokenByAuthorizationCodeAsync(context.Code, new Uri(BaseAddress), credential, GraphResourceId);
```
Looks like it's trying to get a dependency on `System.Runtime.Serialization.Json`, which would be a Framework 4.x assembly, right?
```
System.TypeInitializationException: The type initializer for 'Microsoft.IdentityModel.Clients.ActiveDirectory.Authenticator' threw an exception. ---> System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime.Serialization.Json, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified. ---> System.IO.FileNotFoundException: Could not load the specified file.
 at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyName(AssemblyName assemblyName)
 at System.Runtime.Loader.AssemblyLoadContext.Resolve(IntPtr gchManagedAssemblyLoadContext, AssemblyName assemblyName)
 --- End of inner exception stack trace ---
 at Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticatorTemplate.CreateFromHost(String host)
 at Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticatorTemplateList..ctor()
 at Microsoft.IdentityModel.Clients.ActiveDirectory.Authenticator..cctor()
 --- End of inner exception stack trace ---
 at Microsoft.IdentityModel.Clients.ActiveDirectory.Authenticator.<UpdateFromTemplateAsync>d__0.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Microsoft.IdentityModel.Clients.ActiveDirectory.AcquireTokenHandlerBase.<PreRunAsync>d__6.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Microsoft.IdentityModel.Clients.ActiveDirectory.AcquireTokenHandlerBase.<RunAsync>d__0.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContext.<AcquireTokenByAuthorizationCodeCommonAsync>d__46.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContext.<AcquireTokenByAuthorizationCodeAsync>d__12.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at AzureADTest2.Startup.<>c.<<Configure>b__16_4>d.MoveNext() in C:\PACCHARTER\AzureADTest50\src\AzureADTest50\Startup.cs:line 98
```
One other thing I noticed: When I do a sign off with
```c#
if (Context.User.Identity.IsAuthenticated)
{
Context.Response.HttpContext.Authentication.SignOutAsync(OpenIdConnectAuthenticationDefaults.AuthenticationScheme);
Context.Response.HttpContext.Authentication.SignOutAsync(CookieAuthenticationDefaults.AuthenticationScheme);
}
```
the user is not redirected back to AAD for a full logout. The UI simply reloads. The claims are dropped; however, the user is left in a signed-in state in AAD. If you then sign in again, there is a brief redirect to AAD, but it doesn't ask for credentials ... it just redirects back to the app since the user isn't being signed out fully with the `SignOutAsync` method.

**Comment by kpanwar** [comment]
Yes, ADAL uses a minimum version of .NET 4.5. Whats your version? Also, ADAL does not play any role in sign out flows so I cannot help you there. You could probably look at MSDN forums for help.

**Comment by GuardRex** [comment]
I'm on `dnxcore50`. I do have a fully working `dnx451` app running ... that one works great. I was trying to get a `dnxcore50` version running. The sample app that I reference in https://github.com/aspnet/Home/issues/674#issuecomment-112822724 is located at https://github.com/AzureADSamples/WebApp-WebAPI-OpenIdConnect-AspNet5 with a `project.json` file at https://github.com/AzureADSamples/WebApp-WebAPI-OpenIdConnect-AspNet5/blob/master/TodoListWebApp/project.json shows that it was originally working back before the "big rename" and lists in its framework dependencies `aspnetcore50`. Therefore, I was waiting for a version of the library that would work for `dnxcore50`. That's your team, right? If not, `AcquireTokenByAuthorizationCodeAsync` method is part of `Microsoft.IdentityModel.Clients.ActiveDirectory`, do you know who I would report this issue to?

Oh, yes, I should have checked. I see now that `SignOutAsync` comes from `Microsoft.AspNet.Http.Abstractions`. I'm not sure where to report that issue if it still lingers. I'm only running this app locally against AAD ... it might start working when I put the app on the Azure VM and give the base address and return addresses that are not localhost addresses. I'll hold off on trying to report that issue further until I figure out if it's really an issue and not just cropped up because I'm running against AAD from localhost (but I think it's going to be real, because I've been using localhost and running locally, and this signout issue just cropped up in the last couple of days).

**Comment by kpanwar** [comment]
@brentschmaltz @tushargupta51 @dstrockis , can you please help @GuardRex with the issue around sign out?

**Comment by GuardRex** [comment]
Cool, thanks. I'm going to rename this ... it's more about `dnxcore50` than anything else.

**Comment by tushargupta51** [comment]
@GuardRex Can you share a repo for this? Or is this sample https://github.com/AzureADSamples/WebApp-WebAPI-OpenIdConnect-AspNet5 close enough?

**Comment by GuardRex** [comment]
@kpanwar @tushargupta51  Ok, that's just strange! I thought for sure that I had added
```
 "System.Runtime.Serialization.Json": "4.0.0-beta-23107"
```
last night along with the updated ADAL version and it didn't work out. I added it again just now, and it works fine. It must have been the long day and I was too tired and messed something up (or those Yuengling's that I drank last night are to blame! lol :beer:). 

No, it's all good now, ADAL works perfectly now for `dnxcore50`.

The `SignOutAsync` is still broken. That maps to `Microsoft.AspNet.Http.Abstractions`, so I'm going to open that over at aspnet/Home.

Thanks again guys for setting this up. ADAL is amazing technology! The **best!**
